refactor(transformer): RegExp transform unbox early

overlookmotel opened the pull request:
Unbox the `RegExpLiteral` early to avoid all lookups thereafter going through the indirection of a `Box`.

overlookmotel commented:
This is an experiment. I'm not sure if it'll make any difference.

overlookmotel commented:
No difference in benchmarks, but it's probably not a bad idea anyway.

rzvxa commented on crates/oxc_transformer/src/regexp/mod.rs:
```diff
@@ -113,9 +113,10 @@ impl<'a> Traverse<'a> for RegExp<'a> {
+        let Expression::RegExpLiteral(regexp) = expr else {
             return;
         };
+        let regexp = &mut **regexp;
```
Maybe we can use `DerefMut` here, I always find too many symbols to be confusing compared to this:

```rust
let regexp = regexp.deref_mut();
```

It's just a personal opinion, Though I may as well share it in case you find it tasteful.

overlookmotel replied:
I agree! `&mut **` is rather confusing. But clippy seems to disagree (always ruining the fun, clippy!)

<img width="507" alt="deref" src="https://github.com/user-attachments/assets/5ce16d34-0ce6-44cb-a5eb-6d0b11836d47">

I think what we need is `Box::as_mut`. #5515

overlookmotel replied:
It's now `regexp.as_mut()`. Much better.

overlookmotel committed:
refactor(transformer): RegExp transform unbox early (#5504)

Unbox the `RegExpLiteral` early to avoid all lookups thereafter going through the indirection of a `Box`.
